text/2018-10-25-tikv-client-rust.md

```diff
@@ -0,0 +1,404 @@
+
+# Detailed design
+
+## Supported targets
+
+We will target the `stable` channel of Rust starting in the Rust 2018 edition. We choose to begin with Rust 2018 so we do not need to concern ourselves with an upgrade path.
+
+We will also support the most recent `nightly` version of Rust, but users should not feel the need to reach for nightly unless they are already using it.
+
+## Naming
+
+The [Rust API Guidelines](https://rust-lang-nursery.github.io/api-guidelines/naming.html) do not prescribe any particular crate name convention.
+
+We choose to name the crate `tikv-client`. Choosing to seperate `tikv` and `client` helps potentially unfamiliar users to immediately understand the intent of the package. `tikvclient`, while understandable, is not immediately parsable by a human.
+
+All structures and functions will otherwise follow the [Rust API Guidelines](https://rust-lang-nursery.github.io/api-guidelines/), some of which will be enforced by `clippy`.
+
+## Installation
+
+To utilize the client programmatically, users will be able to add the `tikv-client` crate to their `Cargo.toml`'s dependencies. Then they must use the crate with `use tikv_client;`. Unfortunately due to Rust’s naming conventions this inconsistency is in place.
+
+## Usage
+
+## Two types of APIs
+
+TiKV provides two types of APIs for developers:
+- The Raw Key-Value API
+
+    If your application scenario does not need distributed transactions or MVCC (Multi-Version Concurrency Control) and only needs to guarantee the atomicity towards one key, you can use the Raw Key-Value API.
+
+- The Transactional Key-Value API
+
+    If your application scenario requires distributed ACID transactions and the atomicity of multiple keys within a transaction, you can use the Transactional Key-Value API.
+
+Generally, the Raw Key-Value API has higher throughput and lower latency compared to the Transactional Key-Value API. If distributed ACID transactions are not required, Raw Key-Value API is preferred over Transactional Key-Value API for better performance and ease of use. **Please be aware that these two types of APIs are mutually exclusive. Users should __never__ mix and use these two types of APIs on a __same__ TiKV cluster.**
+
+The client provides two types of APIs in two separate modules for developers to choose from.
+
+### The common data types
+
+- `Key`: raw binary data
+- `Value`: raw binary data
+- `KvPair`: Key-value pair type
+- `KvFuture`: A specialized Future type for TiKV client
+- `Config`: Configuration for client
+
+### Raw Key-Value API Basic Usage
+
+To use the Raw Key-Value API, take the following steps:
+
+1. Create an instance of Config to specify endpoints of PD (Placement Driver) and optional security config
+
+    ```rust
+    let config = Config::new(vec!["127.0.0.1:3379"]).with_security(
+        PathBuf::from("/path/to/ca.pem"),
+        PathBuf::from("/path/to/client.pem"),
+        PathBuf::from("/path/to/client-key.pem"),
```

Comment:
Must we use `PathBuf`s in the security config? Although the config is initialized only once in almost all cases, users might not want to care about the type conversion. What's worse is that this introduces needless copies and allocations when we only have a reference to the path (like the example here).

However, I cannot come up with an alternative elegant enough. The function can receive `AsRef<Path>`s, but we should allow that the types of the parameters are different. Thus, we need three type parameters. It might make the signature of the function ugly. Otherwise, we can ask users to pass `&Path`s in, which looks like the usage her, but require the users to do the conversion themselves.

Reply:
The type of arguments is `impl Into<PathBuf>`, so you can actually write it as `with_security("/path/to/ca.pem", "/path/to/client.pem", "/path/to/client-key.pem")`

Reply:
Ah , I forgot the impl Trait syntax. But why `Into<PathBuf>` instead of `AsRef<Path>`?

Reply:
Config might be cloned and copied to underlying code. Therefore it owns data and PathBuf is what we want.

Reply:
BTW. In case you didn't know. The latest discussion about public api is at [here](https://github.com/tikv/client-rust/pull/10)

Reply:
Really sorry for missing that PR.

Having looked at that PR, I still doubt about the necessity of cloning the content of the config. However, it is reasonable that `Config` might be cloned.

So, I would recommend to read the content of the files on creating `Config` and have `Config` own `Arc`s of the contents. And, will it become clearer to have a separate struct to save the credentials? So it will look like:
```rust
struct Credentials {
    ca: Vec<u8>,
    cert: Vec<u8>,
    key: Vec<u8>,
}

struct Config {
    pd_endpoints: Arc<Vec<String>>,
    credentials: Option<Arc<Credentials>>,
}
```

What do you think?

Reply:
It's probably not a good idea to store key directly in memory.
Talking about Credentials struct, if users don't need to know it why bother adding it?

Reply:
Agree with you. It increases the risk of exposing keys through memory vulnerabilities (like Heartbleed or even Spectre and Meltdown) indeed(, although possibility is decreased by using Rust). 

The `Credentials` struct is not meant to be public to users. I just consider it an tiny improvement on the implementation. With this, we need only one pattern matching for the `Option` without extra `unwrap`s. (So perhaps the discussion about this should be inside of `client-rust` repo instead of this?)

Reply:
Oh, I see what you mean. You can always do this.
```rust
    if let (Some(ca), Some(cert), Some(key)) = (c.ca, c.cert, c.key) {
        println!("sec");
    } else {
        println!("nope");
    }
```

Reply:
I guess implementation detail discussion can go to client-rust repository.

Reply:
Neat solution. No doubts about your design now. Thanks!